File: CS/Classes/GridViewHelper.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Views.Base.ViewInfo;



namespace WindowsApplication1
{
    public class GridViewHelper
    {

        private GridView gridView;
        private GridCell sourceGridCell;
        private GridCell lastMouseDownMoveCell;


        public GridViewHelper(GridView gridView)
        {
            this.gridView = gridView;
            this.gridView.MouseDown += new MouseEventHandler(gridView_MouseDown);
            this.gridView.MouseMove += new MouseEventHandler(gridView_MouseMove);
            this.gridView.MouseUp += new MouseEventHandler(gridView_MouseUp);
            this.gridView.CustomDrawCell += new RowCellCustomDrawEventHandler(gridView_CustomDrawCell); 
        }



        private void gridView_MouseDown(object sender, MouseEventArgs e)
        {
            BaseView view = gridView.GridControl.GetViewAt(e.Location);
            (view as GridView).GridControl.Cursor = Cursors.Default;
            BaseHitInfo baseHI = view.CalcHitInfo(e.Location);
            GridHitInfo gridHI = baseHI as GridHitInfo;

            if (gridHI.HitTest == GridHitTest.RowEdge)
            {
                sourceGridCell = null;
                lastMouseDownMoveCell = null;
                baseHI = view.CalcHitInfo(new Point(e.X, e.Y - 3));
                gridHI = baseHI as GridHitInfo;
                if (gridHI.HitTest == GridHitTest.RowCell)
                {
                    object val = (view as GridView).GetRowCellValue(gridHI.RowHandle, gridHI.Column);
                    if (val == null || val.ToString() == string.Empty) return;
                    sourceGridCell = new GridCell(gridHI.RowHandle, gridHI.Column);
                    lastMouseDownMoveCell = new GridCell(sourceGridCell.RowHandle, sourceGridCell.Column);
                }
            }
        }



        private void gridView_CustomDrawCell(object sender, RowCellCustomDrawEventArgs e)
        {
            if (lastMouseDownMoveCell == null || sourceGridCell == null) return;
            if (e.RowHandle == lastMouseDownMoveCell.RowHandle && e.Column == lastMouseDownMoveCell.Column)
                e.Graphics.DrawRectangle(new Pen(Color.Gray), new Rectangle(e.Bounds.X, e.Bounds.Y, e.Bounds.Width, e.Bounds.Height));
        }



        private void gridView_MouseMove(object sender, MouseEventArgs e)
        {
            BaseView view = gridView.GridControl.GetViewAt(e.Location);
            BaseHitInfo baseHI = view.CalcHitInfo(e.Location);
            GridHitInfo gridHI = baseHI as GridHitInfo;

            if (gridHI.HitTest == GridHitTest.RowCell && sourceGridCell != null)
            {
                bool is_changed = false;

                if (lastMouseDownMoveCell.RowHandle != gridHI.RowHandle || lastMouseDownMoveCell.Column.FieldName != gridHI.Column.FieldName)
                {
                    is_changed = true;
                    int rowHandle = lastMouseDownMoveCell.RowHandle;
                    GridColumn gridCol = lastMouseDownMoveCell.Column;
                    lastMouseDownMoveCell = new GridCell(gridHI.RowHandle, gridHI.Column);
                    (view as GridView).InvalidateRowCell(rowHandle, gridCol);                        
                }

                if (is_changed) (view as GridView).InvalidateRowCell(lastMouseDownMoveCell.RowHandle, lastMouseDownMoveCell.Column);

            }
            else if (gridHI.HitTest == GridHitTest.RowEdge && sourceGridCell == null)
            {
                baseHI = view.CalcHitInfo(new Point(e.X, e.Y - 3));
                gridHI = baseHI as GridHitInfo;
                if (gridHI.HitTest == GridHitTest.RowCell)
                    if (gridHI.RowHandle == (view as GridView).FocusedRowHandle && gridHI.Column == (view as GridView).FocusedColumn)
                       (view as GridView).GridControl.Cursor = Cursors.SizeAll;
            }
            else (view as GridView).GridControl.Cursor = Cursors.Default;
        }



        private void gridView_MouseUp(object sender, MouseEventArgs e)
        {
            if (sourceGridCell != null && lastMouseDownMoveCell != null)
            {
                BaseView view = gridView.GridControl.GetViewAt(e.Location);
                object val = (view as GridView).GetRowCellValue(sourceGridCell.RowHandle, sourceGridCell.Column);
                (view as GridView).SetRowCellValue(sourceGridCell.RowHandle, sourceGridCell.Column, "");
                (view as GridView).SetRowCellValue(lastMouseDownMoveCell.RowHandle, lastMouseDownMoveCell.Column, val);               
                (view as GridView).ClearSelection();
                (view as GridView).SelectCell(lastMouseDownMoveCell.RowHandle, lastMouseDownMoveCell.Column);
                (view as GridView).FocusedRowHandle = lastMouseDownMoveCell.RowHandle;
                (view as GridView).FocusedColumn = lastMouseDownMoveCell.Column;
                sourceGridCell = null;
                int rowHandle = lastMouseDownMoveCell.RowHandle;
                GridColumn gridCol = lastMouseDownMoveCell.Column;
                lastMouseDownMoveCell = null;
                (view as GridView).InvalidateRowCell(rowHandle, gridCol);
            }
        }
    }
}

File: VB/Classes/GridViewHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Views.Base.ViewInfo



Namespace WindowsApplication1
	Public Class GridViewHelper

		Private gridView As GridView
		Private sourceGridCell As GridCell
		Private lastMouseDownMoveCell As GridCell


		Public Sub New(ByVal gridView As GridView)
			Me.gridView = gridView
			AddHandler gridView.MouseDown, AddressOf gridView_MouseDown
			AddHandler gridView.MouseMove, AddressOf gridView_MouseMove
			AddHandler gridView.MouseUp, AddressOf gridView_MouseUp
			AddHandler gridView.CustomDrawCell, AddressOf gridView_CustomDrawCell
		End Sub



		Private Sub gridView_MouseDown(ByVal sender As Object, ByVal e As MouseEventArgs)
			Dim view As BaseView = gridView.GridControl.GetViewAt(e.Location)
			TryCast(view, GridView).GridControl.Cursor = Cursors.Default
			Dim baseHI As BaseHitInfo = view.CalcHitInfo(e.Location)
			Dim gridHI As GridHitInfo = TryCast(baseHI, GridHitInfo)

			If gridHI.HitTest = GridHitTest.RowEdge Then
				sourceGridCell = Nothing
				lastMouseDownMoveCell = Nothing
				baseHI = view.CalcHitInfo(New Point(e.X, e.Y - 3))
				gridHI = TryCast(baseHI, GridHitInfo)
				If gridHI.HitTest = GridHitTest.RowCell Then
					Dim val As Object = (TryCast(view, GridView)).GetRowCellValue(gridHI.RowHandle, gridHI.Column)
					If val Is Nothing OrElse val.ToString() = String.Empty Then
						Return
					End If
                    sourceGridCell = New GridCell(gridHI.RowHandle, gridHI.Column)
                    lastMouseDownMoveCell = New GridCell(sourceGridCell.RowHandle, sourceGridCell.Column)
				End If
			End If
		End Sub



		Private Sub gridView_CustomDrawCell(ByVal sender As Object, ByVal e As RowCellCustomDrawEventArgs)
			If lastMouseDownMoveCell Is Nothing OrElse sourceGridCell Is Nothing Then
				Return
			End If
            If e.RowHandle = lastMouseDownMoveCell.RowHandle AndAlso e.Column.FieldName = lastMouseDownMoveCell.Column.FieldName Then
                e.Graphics.DrawRectangle(New Pen(Color.Gray), New Rectangle(e.Bounds.X, e.Bounds.Y, e.Bounds.Width, e.Bounds.Height))
            End If
		End Sub



		Private Sub gridView_MouseMove(ByVal sender As Object, ByVal e As MouseEventArgs)
			Dim view As BaseView = gridView.GridControl.GetViewAt(e.Location)
			Dim baseHI As BaseHitInfo = view.CalcHitInfo(e.Location)
			Dim gridHI As GridHitInfo = TryCast(baseHI, GridHitInfo)

			If gridHI.HitTest = GridHitTest.RowCell AndAlso sourceGridCell IsNot Nothing Then
				Dim is_changed As Boolean = False

                If lastMouseDownMoveCell.RowHandle <> gridHI.RowHandle OrElse lastMouseDownMoveCell.Column.FieldName <> gridHI.Column.FieldName Then
                    is_changed = True
                    Dim rowHandle As Integer = lastMouseDownMoveCell.RowHandle
                    Dim gridCol As GridColumn = lastMouseDownMoveCell.Column
                    lastMouseDownMoveCell = New GridCell(gridHI.RowHandle, gridHI.Column)
                    TryCast(view, GridView).InvalidateRowCell(rowHandle, gridCol)
                End If


                If is_changed Then
                    TryCast(view, GridView).InvalidateRowCell(lastMouseDownMoveCell.RowHandle, lastMouseDownMoveCell.Column)
                End If

            ElseIf gridHI.HitTest = GridHitTest.RowEdge AndAlso sourceGridCell Is Nothing Then
                baseHI = view.CalcHitInfo(New Point(e.X, e.Y - 3))
                gridHI = TryCast(baseHI, GridHitInfo)
                If gridHI.HitTest = GridHitTest.RowCell Then
                    If gridHI.RowHandle = (TryCast(view, GridView)).FocusedRowHandle AndAlso gridHI.Column Is (TryCast(view, GridView)).FocusedColumn Then
                        TryCast(view, GridView).GridControl.Cursor = Cursors.SizeAll
                    End If
                End If
            Else
                TryCast(view, GridView).GridControl.Cursor = Cursors.Default
            End If
		End Sub



		Private Sub gridView_MouseUp(ByVal sender As Object, ByVal e As MouseEventArgs)
			If sourceGridCell IsNot Nothing AndAlso lastMouseDownMoveCell IsNot Nothing Then
				Dim view As BaseView = gridView.GridControl.GetViewAt(e.Location)
				Dim val As Object = (TryCast(view, GridView)).GetRowCellValue(sourceGridCell.RowHandle, sourceGridCell.Column)
				TryCast(view, GridView).SetRowCellValue(sourceGridCell.RowHandle, sourceGridCell.Column, "")
				TryCast(view, GridView).SetRowCellValue(lastMouseDownMoveCell.RowHandle, lastMouseDownMoveCell.Column, val)
				TryCast(view, GridView).ClearSelection()
				TryCast(view, GridView).SelectCell(lastMouseDownMoveCell.RowHandle, lastMouseDownMoveCell.Column)
				TryCast(view, GridView).FocusedRowHandle = lastMouseDownMoveCell.RowHandle
				TryCast(view, GridView).FocusedColumn = lastMouseDownMoveCell.Column
				sourceGridCell = Nothing
				Dim rowHandle As Integer = lastMouseDownMoveCell.RowHandle
				Dim gridCol As GridColumn = lastMouseDownMoveCell.Column
				lastMouseDownMoveCell = Nothing
				TryCast(view, GridView).InvalidateRowCell(rowHandle, gridCol)
			End If
		End Sub
	End Class
End Namespace
